docs/externalGrading.md

```diff
@@ -163,15 +163,17 @@ The following is an example of a well-structured course layout:
 
 ## Grading results
```

Comment:
Should we also add documentation to `server.py` and/or element-writing docs to say that `grade()` can set `gradable = false`?

Reply:
Since we're no longer parsing, should we still be setting `data['format_errors']`?  Should we have a new key like `data['grading_errors']`?

Reply:
I'm not in love with the name `format_errors`, but I'm tentatively inclined to stick with only having one set of errors that correspond to `gradable = false`. My thinking is that it's sort of an implementation detail in the question about whether the "invalid" status of the submission is detected during `parse()` or `grade()`. We try and detect it during `parse()`, because that's friendlier for the student, but in any case we will treat the outcome the same (that is, `gradable = false` and `format_errors = something`). This has the advantage that we don't need to add separate code paths to handle `grading_errors`.

Thinking about it more, I could even talk myself into liking the name `format_errors`. Certainly for code that doesn't compile, that sounds a lot like a format error of some kind 😄

What do you think? Or is there some advantage to separating `grading_errors` from `format_errors`?

Reply:
I don't think there's an advantage to having it separate &mdash; I just don't know how I feel about the name either.  I suppose it's okay to keep it as-is, just sorta thinking out loud here.

Reply:
Yep, I hear you. I added a couple more comments below. I'd like to be clearer about exactly which properties are being set on the `grading_job` and `submission` when we have valid/invalid submissions. Ideally the rules should be the same whether we discover the invalid status during `parse()` or `grade()`.

Also, it occurs to me that it would be nice to add something about the `grading_job` to the question flow diagram at the bottom of https://prairielearn.readthedocs.io/en/latest/dev-guide/

I think we just need to add information about the `grading_job` to the "Graded submission" and "Ungraded submission" ellipses.